### ISSUE TITLE
iOS Always same appStoreReceipt

### ISSUE THREAD

**Issue by atake831** [j3k0/cordova-plugin-purchase#374]
( Related https://github.com/j3k0/cordova-plugin-purchase/issues/193 )

This issue is about iOS, consumable product.

I have a question the flow of processing about appStoreReceipt.
Because, I am getting back always the same appStoreReceipt.

When we verify appStoreReceipt using store.validator, i think we should implement such as this code.

```
    store.validator = function(product, callback) {
        // we need to refresh appStoreReceipt using storekit.refreshReceipts.
        storekit.refreshReceipts(
            // refresh success
            function(receipt_data){
                // refreshReceipts only updates storekit.appStoreReceipt and localStorage.
                // We need to update product.transaction.appStoreReceipt.
                product.transaction.appStoreReceipt = receipt_data;

                // Request the own server to verify appStoreReceipt.
                store.utils.ajax({
                    url: <: your_server_url :>,
                    method: "POST",
                    data: { "receipt_data": receipt_data },
                    success: function(data) {
                        // success to verify receipt
                    },
                    error: function(status, message) {
                        // fail to verify receipt.
                    }
                });
            },
            // refresh error
            function() {
                 // error handle
            }
      );
    }
```

We cannot use storekit.loadReceipts that is called in prepareForValidation because storekit.loadReceipts does not update appStoreReceipt if storekit.appStoreReceipt already exists.
When user purchase second time, storekit.appStoreReceipt already exists but it is the old one that is for the first purchase. Therefore, we have to use storekit.refreshReceipts.

It is the first time to implement IAP. I need your help. 

Thank you!

**Comment by zsoi** [comment]
I observe this, too. I always get the same receipt for the same transaction all over again.

**Comment by atake831** [comment]
We need to store appStoreReceipt in localStorage to be able to continue purchasing when the app restart in the case that user kill the app before order.finish(). 
But, after order.finish(), appStoreReceipt in localStorage should be deleted. When user purchase again, we get appStoreReceipt that is stored in localStorage. If appStoreReceipt was not deleted, we always get same appStoreReceipt. If deleted, we get new appStoreReceipt by storekit.loadReceipts().

My implementation
```
      store.when(productIdList[i].product_id).verified(function (order) {
          if ( window.localStorage && window.localStorage.sk_appStoreReceipt ) {
              window.localStorage.sk_appStoreReceipt = null;
          }
          if ( storekit.appStoreReceipt ) {
            storekit.appStoreReceipt = null;
          }
          order.finish();
      });
```

It works for me. 
@j3k0 
How about this problem?

**Comment by fattalgazi** [comment]
I have the same issue , any progress ? @atake831 is it ok your solution ?

**Comment by atake831** [comment]
@fattalgazi I'm sorry I missed your message. 
Yes, It works for me well.
So far, no problem.

**Comment by souly1** [comment]
Hi,
I seem to have finally found the solution I was looking for (the above offered work around did not work for me)

Scenario:
Having multiple auto-renew subscriptions, purchasing one and upgrading to the other /purchasing another after first purchase

Buggy result:
Apple's validation would validate both purchases with same appStoreReceipt causing them both to either be valid or invalid.

The pull request solves this issue.

@j3k0 , I suspect the fix solves a bunch of issues here (such as https://github.com/j3k0/cordova-plugin-purchase/issues/473 and https://github.com/j3k0/cordova-plugin-purchase/issues/215) which might not be identified correctly, but the symptom of failure to differentiate between purchases is the same)

Pull request:
https://github.com/j3k0/cordova-plugin-purchase/pull/585

The source for solution:
http://stackoverflow.com/questions/33165384/iap-in-hybrid-app-cordova-itunes-receipt-contains-only-the-first-transaction

**Comment by j3k0** [comment]
Fixed by 4d553aa (according to @souly1)

**Comment by Renzo04** [comment]
Hi, i have a problem with this patch with AutoRenewableSubscriptions, in sandbox environment its worked fine, but in production, the appStoreReceipt is refreshed each time store.refresh() is called, but the information of the receipt is the same, i'm storing the appStoreReceipts in my server and making validations from there with iOS and Android, but now i have multiple appStoreReceipts with the same information...
